resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/HadoopGlobalFeatureDriverStep.scala

```diff
@@ -0,0 +1,151 @@
+    private val isKerberosEnabled = kubeTokenManager.isSecurityEnabled
+
+    require(maybeKeytab.forall( _ => isKerberosEnabled ),
+      "You must enable Kerberos support if you are specifying a Kerberos Keytab")
+
+    require(maybeExistingSecretName.forall( _ => isKerberosEnabled ),
+      "You must enable Kerberos support if you are specifying a Kerberos Secret")
+
+    KubernetesUtils.requireBothOrNeitherDefined(
+      maybeKeytab,
+      maybePrincipal,
+      "If a Kerberos principal is specified you must also specify a Kerberos keytab",
+      "If a Kerberos keytab is specified you must also specify a Kerberos principal")
+
+    KubernetesUtils.requireBothOrNeitherDefined(
+      maybeExistingSecretName,
+      maybeExistingSecretItemKey,
+      "If a secret data item-key where the data of the Kerberos Delegation Token is specified" +
+        " you must also specify the name of the secret",
+      "If a secret storing a Kerberos Delegation Token is specified you must also" +
+        " specify the item-key where the data is stored")
+
+    require(kubernetesConf.hadoopConfDir.isDefined, "Ensure that HADOOP_CONF_DIR is defined")
+    private val hadoopConfDir = kubernetesConf.hadoopConfDir.get
+    private val hadoopConfigurationFiles = kubeTokenManager.getHadoopConfFiles(hadoopConfDir)
+
+    // Either use pre-existing secret or login to create new Secret with DT stored within
+    private val hadoopSpec: Option[KerberosConfigSpec] = (for {
+      secretName <- maybeExistingSecretName
+      secretItemKey <- maybeExistingSecretItemKey
+    } yield {
+      KerberosConfigSpec(
+         dtSecret = None,
+         dtSecretName = secretName,
+         dtSecretItemKey = secretItemKey,
+         jobUserName = kubeTokenManager.getCurrentUser.getShortUserName)
+    }).orElse(
+      if (isKerberosEnabled) {
+         Some(HadoopKerberosLogin.buildSpec(
+             conf,
+             kubernetesConf.appResourceNamePrefix,
+             kubeTokenManager))
+       } else None )
+
+    override def configurePod(pod: SparkPod): SparkPod = {
+      val hadoopBasedSparkPod = HadoopBootstrapUtil.bootstrapHadoopConfDir(
+        hadoopConfDir,
+        kubernetesConf.hadoopConfigMapName,
+        kubeTokenManager,
+        pod)
+      (for {
+        hSpec <- hadoopSpec
+        krb5fileLocation <- kubernetesConf.get(KUBERNETES_KERBEROS_KRB5_FILE)
+      } yield {
+        HadoopBootstrapUtil.bootstrapKerberosPod(
+            hSpec.dtSecretName,
+            hSpec.dtSecretItemKey,
+            hSpec.jobUserName,
+            krb5fileLocation,
+            kubernetesConf.kRBConfigMapName,
+            hadoopBasedSparkPod)
+      }).getOrElse(
+        HadoopBootstrapUtil.bootstrapSparkUserPod(
+          kubeTokenManager.getCurrentUser.getShortUserName,
+          hadoopBasedSparkPod))
+    }
+
+    override def getAdditionalPodSystemProperties(): Map[String, String] = {
+      val resolvedConfValues = hadoopSpec.map{ hSpec =>
```

Comment:
space before `{`

(I thought the style checker complained about that kind of thing...)